oximeter/oximeter/src/test_util.rs

```diff
@@ -104,3 +114,169 @@ pub fn generate_test_samples(
     }
     samples
 }
+
+/// A test function for verifying and possibly updating a target schema.
+///
+/// As schema are updated, the version numbers should change, otherwise the data
+/// can fail to be correctly inserted into the timeseries database. This
+/// function can be used to ensure that.
+///
+/// One can write the expected schema to a file that is checked into the
+/// repository, and then include a test which creates a target and verifies its
+/// schema against the file. If those do not match, this function will fail if
+/// the version number is not greater than the version stored in the file.
+///
+/// Similar to the [`expectorate`][1] crate, users can automatically overwrite
+/// the file by setting the environment variable `EXPECTORATE=overwrite` when
+/// running `cargo test`.
+///
+/// [1]: <https://docs.rs/expectorate/latest/expectorate/>
+pub fn check_target_schema(
+    path: impl AsRef<Path>,
+    target: impl Target,
+) -> anyhow::Result<()> {
+    let actual = serde_json::json!({
+        "name": target.name(),
+        "version": target.version(),
+        "field_names": target
+            .field_names()
+            .into_iter()
+            .map(ToString::to_string)
+            .collect::<Vec<_>>(),
+        "field_types": target
+            .field_types()
+            .into_iter()
+            .map(|f| format!("{f:?}"))
+            .collect::<Vec<_>>(),
+    });
+    check_contents(path, actual)
+}
+
+/// A test function for verifying and possibly updating a metric schema.
+///
+/// As schema are updated, the version numbers should change, otherwise the data
+/// can fail to be correctly inserted into the timeseries database. This
+/// function can be used to ensure that.
+///
+/// One can write the expected schema to a file that is checked into the
+/// repository, and then include a test which creates a metric and verifies its
+/// schema against the file. If those do not match, this function will fail if
+/// the version number is not greater than the version stored in the file.
+///
+/// Similar to the [`expectorate`][1] crate, users can automatically overwrite
+/// the file by setting the environment variable `EXPECTORATE=overwrite` when
+/// running `cargo test`.
+///
+/// [1]: <https://docs.rs/expectorate/latest/expectorate/>
+pub fn check_metric_schema(
+    path: impl AsRef<Path>,
+    metric: impl Metric,
+) -> anyhow::Result<()> {
+    let actual = serde_json::json!({
+        "name": metric.name(),
+        "version": metric.version(),
+        "field_names": metric
+            .field_names()
+            .into_iter()
+            .map(ToString::to_string)
+            .collect::<Vec<_>>(),
+        "field_types": metric
+            .field_types()
+            .into_iter()
+            .map(|f| format!("{f:?}"))
+            .collect::<Vec<_>>(),
+        "datum_type": metric.datum_type(),
+    });
+    check_contents(path, actual)
+}
+
+fn check_contents(
```

Comment:
Any reason to not use the `expectorate` crate itself? Why mimic the behavior?

Reply:
Yeah. The issue is that I don't want to just assert the contents are the same. I want to assert that, if they're different, the version number must have changed. I thought about catching the panic in `expectorate`, but it just seemed complicated. I'll take another look though, I may be splitting hairs.